Add session logic, refactor existing pages to use session service.

eschlange opened the pull request:
* Added necessary controllers, services, and cache.
* Updated eslint config for raw data files.
* Added and updated views for the new functionality.
* Added specs for the new service, controller, and cache.
* Updated home page logic to use new session service.
* Updated configurations to include new files.
* Updated views to utilize new controller logic.

[#95522266]

eschlange commented:
*NOTE* to reviewer: you can ignore the big file, "src/js/constants/SessionContent.js". This file is being imported from REDCap.

ericcf commented on src/js/constants/SessionContent.js:
```diff
@@ -0,0 +1,1304 @@
+var sessionContent = {};
```
Is there a reason this needs to be a global?

eschlange replied:
This was the file given to me by Mark. I'll check with him to see if we can modify the output and formatting of the SessionContent.js file.

ericcf commented on src/js/controllers/SessionsController.js:
```diff
@@ -0,0 +1,79 @@
+
+    this.sessionContent = undefined;
+    this.sessionName = '';
+    this.sessionDescription = 'Click the button to begin...';
```
Why did you decide to store this content on a controller property?

eschlange replied:
This is more of Mark's code. I'll refactor this into the view, since I don't believe that it needs to be dynamic.

ericcf commented on src/js/services/Session.js:
```diff
@@ -0,0 +1,95 @@
+      var session;
+      if (offset > 0) {
+        session = this.CESSATION_SESSION_PRE;
+      } else if (offset <= 0 && offset > -6) {
```
What do you think about creating a named constant for the "magic" value `-6`? I think that would help readability.

eschlange replied:
The constant already existed in the home service, I'll abstract it into the session service and delegate the home service references to the session service.

ericcf commented on src/js/services/SessionQuestion.js:
```diff
@@ -0,0 +1,153 @@
+  'use strict';
+
+  function SessionQuestion() {
+    this.redCapJSONContent = undefined;
```
Why did you choose this name for the property?

eschlange replied:
Mark chose this name, I'll refactor it.

ericcf commented on src/js/services/SessionQuestion.js:
```diff
@@ -0,0 +1,153 @@
+      var self = this;
+      var selected = [];
+
+      $('input[name="' + self.content[this.currentIndex].variable + '"]:checked').each(function() {
```
Is it possible to use `angular.element` so there's not a direct jQuery dependency?

eschlange replied:
I couldn't get it working so I did some searching. I found this article which references the Angular documentation that states that this is not possible without jumping through some hoops: http://stackoverflow.com/a/17230735/512463. Most of the posts recommend including the full jQuery lib if using selectors.

ericcf commented on src/js/services/SessionQuestion.js:
```diff
@@ -0,0 +1,153 @@
+(function() {
```
Are there unit tests for this service? It looks pretty complex.

eschlange commented:
@ericcf I made your recommended changes. I copied in Mark's tests for his parsing service, removed the "magic numbers" and refactored the questionable variable names.

ericcf commented:
I'm getting 34/80 lines covered for the SessionQuestion service. Is that what you have? Do you think that's sufficient and/or is some of the code unnecessary?

eschlange commented:
@ericcf I'm not sure if we'll benefit much from putting more testing into this parser. The session content is not yet finalized.

ericcf commented:
Ok, I'll just write up a separate story for it if you prefer.

eschlange commented:
@ericcf I can write it up. Thanks!
